File: app/src/main/java/com/example/android/newsapp/refactor/usecase/UseCaseFactory.java
package com.example.android.newsapp.refactor.usecase;

import com.example.android.newsapp.refactor.repository.NewsGateway;

public class UseCaseFactory {

    private final NewsGateway gateway;

    public UseCaseFactory(NewsGateway gateway) {
        this.gateway = gateway;
    }

    public GetNewsUseCase getNewsUseCase(DataCallback callback) {
        return new GetNewsUseCase(gateway, callback);
    }
}

File: app/src/main/kotlin/com/example/android/newsapp/usecase/UseCaseFactory.kt
package com.example.android.newsapp.usecase

import com.example.android.newsapp.model.Result
import com.example.android.newsapp.repository.NewsGateway
import com.example.android.newsapp.usecase.base.DataCallback

class UseCaseFactory constructor(private val gateway: NewsGateway) {

    fun getNewsUseCase(callback: DataCallback<List<Result>>): GetNewsUseCase = GetNewsUseCase(gateway, callback)
}
